src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -1082,24 +1070,52 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
     }
 
     pub fn incremental_args(&self, unit: &Unit) -> CargoResult<Vec<String>> {
-        if self.incremental_enabled {
-            if unit.pkg.package_id().source_id().is_path() {
-                // Only enable incremental compilation for sources the user can modify.
-                // For things that change infrequently, non-incremental builds yield
-                // better performance.
-                // (see also https://github.com/rust-lang/cargo/issues/3972)
-                return Ok(vec![format!("-Zincremental={}",
-                                       self.layout(unit.kind).incremental().display())]);
-            } else if unit.profile.codegen_units.is_none() {
-                // For non-incremental builds we set a higher number of
-                // codegen units so we get faster compiles. It's OK to do
-                // so because the user has already opted into slower
-                // runtime code by setting CARGO_INCREMENTAL.
-                return Ok(vec![format!("-Ccodegen-units={}", ::num_cpus::get())]);
-            }
+        // There's a number of ways to configure incremental compilation right
+        // now. In order of descending priority (first is highest priority) we
+        // have:
+        //
+        // * `CARGO_INCREMENTAL` - this is blanket used unconditionally to turn
+        //   on/off incremental compilation for any cargo subcommand. We'll
+        //   respect this if set.
+        // * `build.incremental` - in `.cargo/config` this blanket key can
+        //   globally for a system configure whether incremental compilation is
+        //   enabled. Note that setting this to `true` will not actually affect
+        //   all builds though. For example a `true` value doesn't enable
+        //   release incremental builds, only dev incremental builds. This can
+        //   be useful to globally disable incremental compilation like
+        //   `CARGO_INCREMENTAL`.
+        // * `profile.dev.incremental` - in `Cargo.toml` specific profiles can
+        //   be configured to enable/disable incremental compilation. This can
+        //   be primarily used to disable incremental when buggy for a project.
+        // * Finally, each profile has a default for whether it will enable
+        //   incremental compilation or not. Primarily development profiles
+        //   have it enabled by default while release profiles have it disabled
+        //   by default.
+        let mut global_cfg = self.config.get_bool("build.incremental")?.map(|c| c.val);
+        let profile = match unit.profile.incremental {
+            Incremental::Explicit(v) => v,
+            Incremental::Default(false) => { global_cfg = Some(false); false }
+            Incremental::Default(true) => true,
+        };
+        if !self.incremental_env.unwrap_or(global_cfg.unwrap_or(profile)) {
+            return Ok(Vec::new())
+        }
```

Comment:
This is probably just nitpicking, but I don't find this "nah, let's make global_cfg mutable"  approach easy to follow :) 

May I suggest the following match instead?

```Rust
let incremental = match (self.incremental_env, global.cfg, unit.profile.incremental) {
    (Some(v), _, _) => v,
    (None, Some(false), _) => false,
    (None, _, Incremental::Explicit(v)) | 
    (None, _, Incremental::Default(v)) => v,
};

if !incremental {
    return Ok(Vec::new())
}
```

And looks like we perhaps we can get rid of distinction between `Incremental::Explicit` and `Incremental::Default`?

Reply:
Heh excellent points! Updated now!